examples/features/captures.effekt.md

```diff
@@ -0,0 +1,135 @@
+    def c = new Counter {
+      def tick() = state = state + 1
+      def current() = state
+    }
+    prog {c}
+  }
+
+  // ...
+  with def myCounter = counter(0);
+  ...
+  ```
+  - on term level unbox, box
+  - `at` at type level
+  - capability, resource, region, lifetime, coeffect
+
+## Captures
+
+Functions are second-class in Effekt, but for a good reason. Functions closing over capabilities must not be returned, otherwise it can not be guaranteed that these capabilities are still in scope.
+Thus, we explicitly have to keep track of these captures and ensure they are in scope upon envoking.
+
+Consider the following example where `divide` is being passed a explicit capability:
+
+```
+def divide(n: Int, m: Int) {exc: Exception}: Int =
+  if (m == 0) { exc.throw("uhoh") }
+  else { n / m }
+```
+
+```effekt:repl
+try { 
+  val res = divide(10, 0) {exc}
+  println(res)
+} with exc: Exception {  
+  def throw(msg) = println("ERROR: " ++ show(msg))
+}
+```
+
+What happens if we return the capability?
+
+```
+def example() = try {
+  exc
+} with exc: Exception { 
+  def throw(msg) = panic("ohno")
+}
+```
+
+By directly returning `exc` from the `try` expression, we break the lexical effect handling reasoning since `exc` is only defined within the `try` body.
+For the same reason, returning a closure is not an option either.
+
+For preventing capabilities escaping their scope, Effekt uses captures. In case of the previous example, the return type of `example` is `Exception at {exc}`. This is called
+a boxed computation. A boxed computation may only be used after unboxing it, which is only permitted when all its captures are in scope.
+
+## Boxing
+
+While this may seem like a restriction, this also has upsides. For example, using boxing, you can have first-class functions. You can even ensure that these first-class functions
+only use a certain set of effects.
+
+Consider a simple scheduler that runs two functions in parallel:
+
+```
+def parallelWrong { f: () => Unit } { g: () => Unit }: Unit = <>
+```
+
+Since `parallelWrong` expects two arbitrary blocks as arguments, these blocks may capture arbitrary capabilities.
+Running both in parallel can have non-deterministic side-effects or introduce data races.
+Thus, we need to enforce that both `f` and `g` are pure, that is, do not capture any capabilities.
+
+```
+def parallel (f: () => Unit at {}, g: () => Unit at {}): Unit = <>
+```
+
+Converting second-class computations to first-class values can be done either explicitly by `box`ing them or implicitly when
+- they occur in return position,
+- on the right-hand side of assignments,
+- or being passed as arguments where first-class values are expected.
+
+```effekt:repl
+parallel(
+  box { () => println("Hello, ") },
+  box { () => println("world!") }
+)
+```
+
+Here both arguments have the capture set `{io}` while the expected capture set is expected to be empty `{}`. Hence, this call does not type-check.
```

Comment:
```suggestion
## Boxing
Only values (which are "first-class") can be returned from computation, be passed as arguments to a datatype constructor, or stored in a mutable reference.

Does this mean that computation (which is "second-class") like functions, capabilities, or objects cannot be returned or stored?

As illustrated above, allowing this without any restriction would lead to unsafety and potential runtime errors. 
Effekt thus takes the following approach: 

> Computation can be packaged up into inert values by **boxing**.
> Computation can be extracted (and reactivated) from boxed values by **unboxing**.

In our example, this means that we can turn the capability `exc` into a value by using the term-level syntax `box c`.

```
def example() = try {
  val cap = box exc;
  println("done")
} with exc: Exception { 
  def throw(msg) = panic("ohno")
}
```

Hovering over cap shows the inferred type is `Exception at {exc}`. While `Exception` is an interface and lives in the universe of computation types, `Exception at {exc}` is a "boxed" type and lives in the universe of value types. 
To prevent capabilities escaping their scope, Effekt tracks their capture in the type of boxed computations. 
In case of our example, the type of `cap` expresses that the capability of type `Exception` can only be used when `exc` is in scope.

TODO continue...

While this may seem like a restriction, this also has upsides. For example, using boxing, you can have first-class functions. You can even ensure that these first-class functions
only use a certain set of effects.

Consider a simple scheduler that runs two functions in parallel:

```
def parallelWrong { f: () => Unit } { g: () => Unit }: Unit = <>
```

Since `parallelWrong` expects two arbitrary blocks as arguments, these blocks may capture arbitrary capabilities.
Running both in parallel can have non-deterministic side-effects or introduce data races.
Thus, we need to enforce that both `f` and `g` are pure, that is, do not capture any capabilities.

```
def parallel (f: () => Unit at {}, g: () => Unit at {}): Unit = <>
```

Converting second-class computations to first-class values can be done either explicitly by `box`ing them or implicitly when
- they occur in return position,
- on the right-hand side of assignments,
- or being passed as arguments where first-class values are expected.

```effekt:repl
parallel(
  box { () => println("Hello, ") },
  box { () => println("world!") }
)
```

Here both arguments have the capture set `{io}` while the expected capture set is expected to be empty `{}`. Hence, this call does not type-check.
```

Reply:
I've fixed the formatting of the suggestion, it should be something like this if you want nested code blocks:
`````
````suggestion
words here
```effekt
code here
```
more words here
````
`````